### DISCUSSION TITLE
Steam Deck Ground Station

### DISCUSSION THREAD

**Discussion by knuckles904** [bitcraze/discussions#257] [Ideas]
This is very much a niche within a niche, but if anyone at Bitcraze happens to be a PC game aficionado who purchased a Steam Deck (handheld PC-think Nintendo Switch form factor but Linux computer), I'd love to see a tutorial on how to set up the Crazyflie client for a Steam Deck. I'm making an attempt but haven't gotten it working yet. The setup seems like a dream come true!

**Comment by knuckles904** [comment]
Edit - I think most of the generic linux instructions work fine but currently trying to get the CF radio connected, and get the CF Client to find the Steam Deck's sticks/buttons

**Comment by knmcguire** [comment]
Oh yes that would be very awesome. We didn't order one as the waiting times seemed to be quite long at that time... but now it seems to be okay again. I'll discuss it internally to see what the others here think!

In the mean time, I saw this video about somebody putting ROS on the steamdeck: https://youtu.be/KbF4r7yCCds. Perhaps somewhere in there could be some pointers on getting the joystick inputs?

**Comment by knuckles904** [comment]
Thanks! That video was helpful actually. 

I'm quite happy to report that cfclient can be installed via "python3 -m pip install cfclient" once miniconda ([miniconda3 Linux 64-bit](https://docs.conda.io/en/latest/miniconda.html)) has been installed on the steam deck (and confirmed that that PATH has been added, was automatically for me) . Removing the readonly config from the Steam Deck was not necessary for me.

Most of the steam deck's controls are passed as a virtual xbox360 controller in Arch linux, so happily that's what CFClient sees! Gyro, touch pads, and back buttons don't map at all, and D-pad maps as 2 digital digital axes instead of 4 buttons. Still plenty enough inputs to use.

I did not have to mess with udev permissions curiously enough; the crazyradio worked fine once CFClient was running.

Now only remains to be seen if this type of install gets wiped out when SteamOS receives an update.

**Reply by knmcguire** [reply]
awesome! Keep us up to date for it. In the mean time, we have put ourselves in the waiting list 😄

**Comment by knmcguire** [comment]
Do you have any update? I had somebody that had questions about the controls on the steamdeck as well, so wondered if you were able to use those at all.

**Reply by knuckles904** [reply]
Yes, for sure. All the main controls show up as a virtual xbox360 controller in Arch linux so CFClient finds them just fine. I'll try to do a full writeup for the whole install and set-up process (not all that long).

**Comment by knuckles904** [comment]
OK, here's my full tutorial for getting CF Client set up on the Steam Deck, at least as of SteamOS 3.3.1

0. Materials needed - Will need a USB C to USB A adapter to plug in the CrazyRadio to the Steam Deck's USB C port (or alternatively a USB C dock with USB A plug). 
You may also want a USB C or Bluetooth keyboard but I installed everything without just using the on-screen keyboard, its just more tedious...
1. Switch to Desktop mode if in gaming mode (Hardware Steam button->Power-Switch to Desktop)
2. Use web browser (firefox by default) to download the latest miniconda installer ("Miniconda 3 Linux 64-bit", not aarch64)
[https://docs.conda.io/en/main/miniconda.html#latest-miniconda-installer-links](https://docs.conda.io/en/main/miniconda.html#latest-miniconda-installer-links)
3. Follow the Linux miniconda installation instructions for miniconda (Konsole is the name for terminal on steam deck's version of linux)
[https://conda.io/projects/conda/en/latest/user-guide/install/linux.html](https://conda.io/projects/conda/en/latest/user-guide/install/linux.html)
For me, with miniconda installer downloaded to the desktop, it was entering into Konsole "bash /home/deck/Desktop/Miniconda3-latest-Linux-x86_64.sh"
4. Confirm Python was successfully installed by entering "python3" into Konsole and verifying you receive ">>>" meaning python interpreter has started. 
If successful, exit python via "quit()"
5. Use the linux (not debain/ubuntu) install instructions at https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/installation/install/
My process was:
5a. Maybe not necessary but go ahead and upgrade pip to get most recent stuff "pip3 install --upgrade pip"
5b. Install CF Client via "pip3 install cfclient"
6. Run CF Client via "python3 /home/deck/enter/bin/cfclient"
7. Assuming CF Client popped up fine, configure the controls as normal in CF Client. Controls for the steam deck should show up automatically as "Microsoft X-Box 360 pad 0". 

Note-Most of the deck's controls work great! Split-axis Yaw on the analog triggers even works. However the gyroscope, touch pads, and 4 extra backside buttons don't map at all, and D-pad maps as 2 digital digital axes instead of 4 buttons (which is different from how a real XBox controller behaves). Still more than enough inputs to use all the functions Crazyflie has.

Troubleshooting-If for some reason you're not seeing the virtual Xbox 360 controller in CFClient, double check that the Steam Deck controls are showing up via Settings->System Settings->Input Devices->Game Controller. You can also calibrate controls in this menu. It could be that Steam Client has been closed, which changes the control/mapping scheme in Desktop Mode. Try opening and minimizing Steam Client in desktop mode to see if this fixes not seeing the virtual XBox controller. 

I did not have to mess with Udev permissions curiously enough; the Crazyradio worked just fine once CFClient was running. Your mileage may vary though.

Lastly, if you previously installed cfclient and want to update to a new release, enter into Konsole "pip3 install cfclient --upgrade"

**Reply by krichardsson** [reply]
Thanks for the detailed tutorial!

**Reply by knmcguire** [reply]
Awesome! Thanks indeed

**Comment by knmcguire** [comment]
Just a final question, you didn't reinstall anything in terms of OS right? You used the OS that already came with the steamdeck?

**Reply by knuckles904** [reply]
That's correct. The deck ships with "SteamOS" which is a customized version of Arch Linux. The desktop mode that I mentioned is basically KDE Plasma, possibly with some tweaks/other stuff installed. 

Outdated but helpful article: [https://arstechnica.com/gaming/2022/02/linux-on-steam-deck-what-you-need-to-know-what-currently-works/](https://arstechnica.com/gaming/2022/02/linux-on-steam-deck-what-you-need-to-know-what-currently-works/)

**Reply by knmcguire** [reply]
noted! thanks 😄
